src/sage/knots/free_knotinfo_monoid.py

```diff
@@ -0,0 +1,479 @@
+            sage: FKIM =  FreeKnotInfoMonoid()
+            sage: K = KnotInfo.K5_1.link().mirror_image()
+            sage: FKIM(K)
+            KnotInfo['K5_1m']
+        """
+        if isinstance(x, tuple):
+            if len(x) == 2:
+                ki, sym = x
+                from sage.knots.knotinfo import KnotInfoBase
+                if isinstance(ki, KnotInfoBase) and isinstance(sym, SymmetryMutant):
+                    mcr = ki.crossing_number()
+                    if mcr > self._max_crossing_number:
+                        self._set_index_dictionary(max_crossing_number=mcr)
+
+                    sym_min = min([sym] + sym.matches(ki))
+                    return self.gen(self._from_knotinfo(ki, sym_min))
+
+        from sage.knots.knot import Knot
+        from sage.knots.link import Link
+        if not isinstance(x, Knot):
+            if isinstance(x, Link):
+                x = Knot(x.pd_code())
+        if isinstance(x, Knot):
+            return self.from_knot(x)
+        return self.element_class(self, x)
+
+    @cached_method
+    def _check_elements(self, knot, elems):
+        r"""
+        Return a matching item from the list in ``elems`` if it exists.
+        Elsewise return ``None``. This is a helper method for .meth:`from_knot`.
+
+        INPUT:
+
+        - ``knot`` -- an instance of :class:`~sage.knots.knot.Knot`
+        - ``elems`` -- a tuple of elements of ``self``
+
+        EXAMPLES::
+
+            sage: from sage.knots.free_knotinfo_monoid import FreeKnotInfoMonoid
+            sage: FKIM =  FreeKnotInfoMonoid()
+            sage: FKIM.inject_variables(select=3)
+            Defining K3_1
+            Defining K3_1m
+            sage: elems = (K3_1, K3_1m)
+            sage: K = Knots().from_table(3, 1)
+            sage: FKIM._check_elements(K, elems)
+            KnotInfo['K3_1m']
+            sage: K = Knots().from_table(4, 1)
+            sage: FKIM._check_elements(K, elems) is None
+            True
+        """
+        for e in elems:
+            k = e.as_knot()
+            if knot.pd_code() == k.pd_code():
+                return e
+            if knot._markov_move_cmp(k.braid()):
+                return e
+        return None
+
+    @cached_method
+    def _from_knot(self, knot):
+        """
+        Create a tuple of element of this abelian monoid which possibly
+        represent ``knot``. This method caches the performance relevant
+        part of :meth:`from_knot`.
+
+        INPUT:
+
+        - ``knot`` -- an instance of :class:`~sage.knots.knot.Knot`
+
+        EXAMPLES::
+
+            sage: from sage.knots.free_knotinfo_monoid import FreeKnotInfoMonoid
+            sage: FKIM =  FreeKnotInfoMonoid()
+            sage: K = KnotInfo.K5_1.link().mirror_image()
+            sage: FKIM._from_knot(K)
+            (KnotInfo['K5_1m'],)
+        """
+        from sage.knots.knotinfo import KnotInfo
+
+        def search_composition(max_cr, hpoly):
```

Comment:
I think this would be better as a (hidden) method or function.

Reply:
This is fixed by the current commit.